### PR TITLE
WIP - triangular distribution

### PR DESCRIPTION
**Pull request by jimenofonseca**
This PR solves issue #1578. Tests in PyMC3 need to pass first.

### CODE REVIEW

**Comment by springcoil** [comment]
Where's the tests? - Adding WIP to flag this as it needs a test to be added.

File: pymc3/distributions/continuous.py [hunk 1]
```diff
@@ -1410,3 +1410,45 @@ def logp(self, value):
...
+
+    def logp(self, value):
+        c = self.c
+        lower = self.lower
+        upper = self.upper
+        return tt.switch(alltrue_elemwise([lower <= value, value < c]),
+                         tt.log(2 * (value - lower) / ((upper - lower) * (c - lower))),
+                         tt.switch(alltrue_elemwise([value == c]), tt.log(2 / (upper - lower)),
+                                   tt.switch(alltrue_elemwise([c < value, value <= upper]),
+                                             tt.log(2 * (upper - value) / ((upper - lower) * (upper - c))), np.inf)))
```
Comment (twiecki) [inline pymc3/distributions/continuous.py]: New new-line.

File: pymc3/distributions/continuous.py [hunk 2]
```diff
@@ -1410,3 +1410,45 @@ def logp(self, value):
             + (-tau * (value - mu)**2
             + tt.log(tau / np.pi / 2.)) / 2.,
             tau > 0, sd > 0)
+
+
+class Triangular(Continuous):
+    """
```
Comment (twiecki) [inline pymc3/distributions/continuous.py:1309]: Doc-string could be improved, see the other doc-strings that include the support of the parameters as well a latex string of the math.

**Comment by jimenofonseca** [comment]
@springcoil sorry, i am new in helping in this repository. What are the tests that should be included? where do i find them and how do i run them?

**Comment by ColCarroll** [comment]
hey @JIMENOFONSECA -- check out `pymc3/tests/test_distributions.py` and `pymc3/tests/test_distributions/random.py`.  We enforce there that `pymc3` is [statistically] as good as `scipy`, meaning you'll compare the distribution to [scipy.stats.triang](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.triang.html).  It should look a lot like the `test_normal` and `TestNormal` cases.  Let me know if that's still confusing!

To run the tests, you can either use docker (see docs!), or assuming everything is where it should be, `./scripts/test.sh -vx pymc3/tests/test_distributions.py:TestMatchesScipy` should also work locally.

**Comment by jimenofonseca** [comment]
@ColCarroll thanks, the test runs now, I hope this is the way to use it. please check it out.

 @twiecki I am not used to Latec and the structure of the PyMC3 documentation . Would you mind improving the Doc-string?

File: pymc3/distributions/__init__.py [hunk 3]
```diff
@@ -24,6 +24,7 @@
 from .continuous import ExGaussian
 from .continuous import VonMises
 from .continuous import SkewNormal
+from .continuous import Triangular
```
Comment (ColCarroll) [inline pymc3/distributions/__init__.py:26]: you'll need to add `'Triangular'` to the `__all__` list in this file as well -- should fix the test error.

**Comment by twiecki** [comment]
@JIMENOFONSECA I resolved a merge conflict, make sure to pull.

File: pymc3/distributions/__init__.py (hunk 3 above)
Comment (jimenofonseca) [reply pymc3/distributions/__init__.py:26]: done!, thank you

**Comment by jimenofonseca** [comment]
@ColCarroll @twiecki so i guess it is ready now? :-)

**Comment by twiecki** [comment]
thanks @JIMENOFONSECA!

**Comment by ColCarroll** [comment]
👍

**Comment by ferrine** [comment]
Hmm, maybe I'm writing too late. I see there is no transform for Triangular distribution, Is it ok?

**Comment by twiecki** [comment]
@ferrine you're right, good catch.

**Comment by ferrine** [comment]
Docstring can be better also

**Comment by twiecki** [comment]
@JIMENOFONSECA can you add those?

**Comment by jimenofonseca** [comment]
Sorry, what is transform of Triangular?


On 25 Jan 2017, at 22:57, Thomas Wiecki <notifications@github.com<mailto:notifications@github.com>> wrote:


@JIMENOFONSECA<https://github.com/JIMENOFONSECA> can you add those?

**Comment by ferrine** [comment]
@JIMENOFONSECA I thing the same as for [Uniform](https://github.com/architecture-building-systems/pymc3/blob/833d1cb45df6a113c757d3a0ba27a5559b39eb67/pymc3/distributions/continuous.py#L105) will fit there

**Comment by jimenofonseca** [comment]
You mean the doc string?, if so i do not do latec and will need somebody to contribute this part i can provide the equations in python language..

On 25 Jan 2017, at 23:12, Maxim Kochurov <notifications@github.com<mailto:notifications@github.com>> wrote:


@JIMENOFONSECA<https://github.com/JIMENOFONSECA> I thing the same as for Uniform<https://github.com/architecture-building-systems/pymc3/blob/833d1cb45df6a113c757d3a0ba27a5559b39eb67/pymc3/distributions/continuous.py#L105> will fit there

**Comment by twiecki** [comment]
OK, but please add the transform, which you can gather from the Uniform.

**Comment by srihari4mbatech** [comment]
Hi,

If I upgrade Pymc3, can I get Triangular distribution, in it?

**Comment by twiecki** [comment]
Yes, if you install the master.

**Pull request merged**